src/raftstore/store/snap.rs

```diff
@@ -0,0 +1,284 @@
+        Ok(written)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        if self.tmp_file.is_none() {
+            return Ok(());
+        }
+        self.tmp_file.as_mut().unwrap().0.flush()
+    }
+}
+
+impl Drop for SnapFile {
+    fn drop(&mut self) {
+        if let Some((_, path)) = self.tmp_file.take() {
+            if let Err(e) = fs::remove_file(&path) {
+                warn!("failed to delete temporary file {}: {:?}", path, e);
+            }
+        }
+    }
+}
+
+/// `SnapManagerCore` trace all current processing snapshots.
+pub struct SnapManagerCore {
+    // directory to store snapfile.
+    base: String,
+    // bool indicate if the snapshot is being generated by the peer on current store.
+    // usize indicate the times the snapkey has been registered.
+    registry: HashMap<SnapKey, (bool, usize)>,
+}
+
+impl SnapManagerCore {
+    pub fn new<T: Into<String>>(path: T) -> SnapManagerCore {
+        SnapManagerCore {
+            base: path.into(),
+            registry: map![],
+        }
+    }
+
+    pub fn try_recover(&self) -> io::Result<()> {
+        let path = Path::new(&self.base);
+        if !path.exists() {
+            try!(fs::create_dir_all(path));
+        }
+        if !path.is_dir() {
+            return Err(io::Error::new(ErrorKind::Other,
+                                      format!("{} should be a directory", path.display())));
+        }
+        for path in try!(fs::read_dir(path)) {
+            let p = try!(path);
+            if !try!(p.file_type()).is_file() {
+                continue;
+            }
+            try!(fs::remove_file(p.path()));
+            // TODO: resume applying when suitable
+        }
+        Ok(())
+    }
+
+    #[inline]
+    pub fn get_snap_file(&self, key: &SnapKey, is_sending: bool) -> io::Result<SnapFile> {
+        SnapFile::new(&self.base, is_sending, key)
+    }
+
+    pub fn register(&mut self, key: SnapKey, is_sending: bool) {
+        debug!("register [key: {}, is_sending: {}]", key, is_sending);
+        match self.registry.entry(key) {
+            Entry::Occupied(mut e) => {
+                if e.get().0 == is_sending {
+                    e.get_mut().1 += 1;
+                } else {
+                    info!("seems leadership changed, cleanup old snapfiles");
+                    if let Ok(f) = SnapFile::new(&self.base, !is_sending, e.key()) {
+                        f.delete();
+                    }
+                    e.insert((is_sending, 1));
+                }
+            }
+            Entry::Vacant(e) => {
+                e.insert((is_sending, 1));
+            }
+        }
+    }
+
+    pub fn deregister(&mut self, key: &SnapKey, is_sending: bool) {
+        debug!("deregister [key: {}, is_sending: {}]", key, is_sending);
+        let mut need_cleanup = false;
+        match self.registry.get_mut(key) {
+            Some(&mut (b, ref mut cnt)) if b == is_sending => {
+                *cnt -= 1;
```

Comment:
No, if is_sending doesn't match, nothing can be done. It has probably been handled when registered with a different is_sending.